src/win32/waiter.cpp

```diff
@@ -135,6 +144,15 @@ int WinWaiter::wait()
 bool WinWaiter::addhandle(HANDLE handle, int flag)
 {
     assert(handles.size() == flags.size() && handles.size() >= index);
+
+#ifdef DEBUG
+    for (unsigned i = index; i--; )
+    {
+        // double check we only add one of each handle
+        assert(handles[i] != handle);
+    }
```

Comment:
Perhaps `foreach()` is more appropriate for these cases. But it probably depends on the each person :wink:

Reply:
It is a vector, but for performance (because it kept showing up in the profiler, it was measured and not just guesswork), we don't keep modifying the size of the vector  (even just changing the size with resize() and push_back()).  The size of the vector is the maximum number of handles we've ever waited on.    With this PR that number will be a lot smaller, and perhaps not such a performance issue, but having already investigated and fixed that I'd like to keep the benefit.   So we just use index to keep track of how many of the vector's entries we have filled in for this upcoming wait.   So if we used `for(auto h: handles)` , we would iterate over too many entries and possibly get the wrong answer - I do prefer that form too, when appropriate.  thanks

Reply:
Oh, now I see why to use the `index`. I should have checked how it's set in advance. My apologies, since it makes perfect sense :wink: